a- Read	the	iris	dataset	from	the	following	URL:	
'https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv'			
and	assign	it	to	a	Pandas	DataFrame	as	you	learned	in	tutorial	Lab2-3.	

In [116]:
from sklearn.datasets import load_iris 
from sklearn.neighbors import KNeighborsClassifier
from sklearn.cross_validation import train_test_split
from sklearn.metrics import accuracy_score 
import pandas as pd
df = pd.DataFrame()
iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')

def categorical_to_numeric(x):
    if x == 'setosa':
        return 0
    elif x == 'versicolor':
        return 1
    elif x == 'virginica':
        return 2
    

b- Split	 the	 dataset	 into	 testing	 and	 training	 sets	 with	 the	 following	 parameters:	
test_size=0.4,	random_state=1	

In [117]:
iris = load_iris()
X= iris.data
y=iris.target
X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)



c- Instantiate	a	KNN	object	with	K=3,	train	it	on	the	training	set	and	test	it	on	the	testing	set.	
Then,	calculate	the	accuracy	of	your	prediction	as	you	learned	in	Lab3.

In [118]:
K=3
knn = KNeighborsClassifier(n_neighbors=K)
knn.fit(X_train, y_train)
y_predict = knn.predict(X_test)
accuracy = accuracy_score(y_test, y_predict)

print(accuracy)

0.983333333333


d- Repeat	part	(c)		for	K=1,	K=5,	K=7,	K=15,	K=28,	K=60	(you	can	simply	write	a	“for	loop”	on	
a	list	with	these	values	to	repeat	the	process	for	you,	and	save	the	final	accuracy	results	
in	a	list	(array)!).	Does	the	accuracy	always	get	better	by	increasing	K?		

In [119]:
X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)    
kList = [1, 5, 7, 15, 28, 60]
accuracyList=[]
for k in kList:
    knn= KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
    accuracyList.append(accuracy)
    
print(accuracyList)

[0.96666666666666667, 0.98333333333333328, 0.98333333333333328, 0.98333333333333328, 0.94999999999999996, 0.80000000000000004]


Does the accuracy always get better by increasing K? 

### It does not look like it's getting better when K=60 its 0.80000000000000004 and  K = 5 is 0.98333333333333328

e- Now,	suppose	that	we	would	like	to	make	prediction	based	on	only	ONE	single	feature.	
To	find	the	best	single	feature,	we	have	to	try	every	feature	individually.	In	other	word,	
we	 have	 to	 repeat	 part	 (c)	 4	 times	 (each	 time	 using	 only	 one	 of	 the	 4	 features),	 and	
compute	the	accuracy	each	time.	Then,	check	the	accuracies.	Which	individual	feature	
provide	 the	 best	 accuracy	 (the	 best	 feature)?	 	 Which	 one	 is	 the	 second	 best	 feature?	
(Note:	you	have	to	train,	test,	and	evaluate	your	model	4	times,	each	time	on	a	dataset	
including	only	one	of	the	features.	You	may	want	to	write	a	“for	loop”	to	generate	the	
datasets,	train,	test,	and	compute	the	accuracy	each	time,	and	save	the	final	accuracy	
results	in	a	list).	

In [121]:
iris_df['label'] = iris_df['species'].apply(categorical_to_numeric)
feature_cols = ['sepal_length','sepal_width','petal_length','petal_width']
# iris_df[0::10]

K=3
knn = KNeighborsClassifier(n_neighbors=K)

oneFeatureList=[]
count = 0

for fc in feature_cols:
    X= iris_df[[fc]]
    y= iris_df['label']
    X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)

    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
    oneFeatureList.append(accuracy)
    print feature_cols[count] + " = " + str(oneFeatureList[count])
    count+=1
    
    
second_best_feature = sorted(oneFeatureList)[-2]
best_feature = sorted(oneFeatureList)[-1]
print ("\nbest feature is : " + str(best_feature))
print ("second best feature is : " + str(second_best_feature) + "\n")
print (oneFeatureList)





sepal_length = 0.633333333333
sepal_width = 0.533333333333
petal_length = 0.966666666667
petal_width = 0.933333333333

best feature is : 0.966666666667
second best feature is : 0.933333333333

[0.6333333333333333, 0.53333333333333333, 0.96666666666666667, 0.93333333333333335]


### The best feature is petal_length = 0.966666666667 and second best feature is petal_width = 0.933333333333

f- Now,	we	want	to	repeat	part	(e),	this	time	using	two	features.	you	have	to	train,	test,	and	
evaluate	 your	 model	 for	 6	 different	 cases:	 using	 (1 st 	 and	 2 nd 	 features),	 (1 st 	 and	 3 rd 		
features),	(1 st 	and	4 th 		features),	(2 nd 		and	3 rd 		features),	(2 nd 	and	4 th 	features),	(3 rd 	and	4 th 		
features)!																																												
Which “feature	pair”	provide	the	best	accuracy?	

In [182]:
iris_df['label'] = iris_df['species'].apply(categorical_to_numeric)
feature_cols=[['sepal_length','sepal_width'],['sepal_length','petal_length'],['sepal_length','petal_width'],['sepal_width','petal_length'],['sepal_width','petal_width'],['petal_length','petal_width']]

K=3
knn = KNeighborsClassifier(n_neighbors=K)
featureList=[]
count = 0
best_accuracy_pair = ""
best_accuracy_value = 0
for fc in feature_cols:
    X= iris_df[fc]
    y= iris_df['label']
    X_train, X_test,y_train,y_test = train_test_split(X, y, test_size=0.4, random_state=1)
    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)
    accuracy = accuracy_score(y_test, y_predict)
    featureList.append(accuracy)
    string = ' and '.join(map(str,feature_cols[count]) )
    print (string + " is " + str(featureList[count]))

    if featureList[count] > best_accuracy_value:
        best_accuracy_value = featureList[count]
        best_accuracy_pair = string
    count += 1
    
print("\n")
print(featureList)
print ("\n The best feature pair accuracy is : " +  best_accuracy_pair + " " + str(best_accuracy_value)  )

sepal_length and sepal_width is 0.65
sepal_length and petal_length is 0.933333333333
sepal_length and petal_width is 0.983333333333
sepal_width and petal_length is 0.95
sepal_width and petal_width is 0.966666666667
petal_length and petal_width is 0.966666666667


[0.65000000000000002, 0.93333333333333335, 0.98333333333333328, 0.94999999999999996, 0.96666666666666667, 0.96666666666666667]

 The best feature pair accuracy is : sepal_length and petal_width 0.983333333333


### sepal length and petal width is the best feature pair 

g- Big	Question:	Does	the	“best	feature	pair”	from	part	(f)	contain	of	both	“first	best	feature”	
and	“second	best	feature”	from	part	(e)?		In	other	word,	can	we	conclude	that	the	“best	
two	features”	for	classification	are	the	“first	best	individual	feature”	and	the	“second	best	
individual	feature”	together?	

### No Because the best two features is petal length and petal width on part e.  But on part f, the best feature pair is sepal lenght and petal width.  

h- Optional	Question:	Justify	your	answer	to	part	(g)!		If	yes,	why?		If	no,	why	not?	

### It is because more data is being analyzed as a dataframe set and what the best feature on just one part of the feature might be offset if used with another set of the feature.  The sum of the second pair could offset the value of the output as a whole.  